applications/IgaApplication/custom_conditions/support_penalty_condition.cpp

```diff
@@ -0,0 +1,176 @@
+                noalias(rLeftHandSideMatrix) += prod(trans(H), H)
+                    * penalty * integration_weight * determinat_jacobian;
+            }
+            if (CalculateResidualVectorFlag) {
+
+                const array_1d<double, 3>& displacement = Has(DISPLACEMENT)
+                    ? this->GetValue(DISPLACEMENT)
+                    : ZeroVector(3);
+
+                Vector u(mat_size);
+                for (IndexType i = 0; i < number_of_nodes; i++)
+                {
+                    const array_1d<double, 3> disp = r_geometry[i].FastGetSolutionStepValue(DISPLACEMENT);
+                    IndexType index = 3 * i;
+                    u[index] = (disp[0] - displacement[0]);
+                    u[index + 1] = (disp[1] - displacement[1]);
+                    u[index + 2] = (disp[2] - displacement[2]);
+                }
+
+                noalias(rRightHandSideVector) -= prod(prod(trans(H), H), u)
+                    * penalty * integration_weight * determinat_jacobian;
+            }
+        }
+        KRATOS_CATCH("")
+    }
+
+    void SupportPenaltyCondition::DeterminantOfJacobianInitial(
+        const GeometryType& rGeometry,
+        Vector& rDeterminantOfJacobian)
+    {
+        const IndexType nb_integration_points = rGeometry.IntegrationPointsNumber();
+        if (rDeterminantOfJacobian.size() != nb_integration_points) {
+            rDeterminantOfJacobian.resize(nb_integration_points, false);
+        }
+
+        const SizeType working_space_dimension = rGeometry.WorkingSpaceDimension();
+        const SizeType local_space_dimension = rGeometry.LocalSpaceDimension();
+        const SizeType nb_nodes = rGeometry.PointsNumber();
+
+        Matrix J = ZeroMatrix(working_space_dimension, local_space_dimension);
+        for (IndexType pnt = 0; pnt < nb_integration_points; pnt++)
+        {
+            const Matrix& r_DN_De = rGeometry.ShapeFunctionsLocalGradients()[pnt];
+            J.clear();
+            for (IndexType i = 0; i < nb_nodes; ++i) {
+                const array_1d<double, 3>& r_coordinates = rGeometry[i].GetInitialPosition();
+                for (IndexType k = 0; k < working_space_dimension; ++k) {
+                    const double value = r_coordinates[k];
+                    for (IndexType m = 0; m < local_space_dimension; ++m) {
+                        J(k, m) += value * r_DN_De(i, m);
+                    }
+                }
+            }
+
+            //Compute the tangent and  the normal to the boundary vector
+            array_1d<double, 3> local_tangent;
+            GetGeometry().Calculate(LOCAL_TANGENT, local_tangent);
+
+            array_1d<double, 3> a_1 = column(J, 0);
+            array_1d<double, 3> a_2 = column(J, 1);
+
+            rDeterminantOfJacobian[pnt] = norm_2(a_1 * local_tangent[0] + a_2 * local_tangent[1]);
+        }
+    }
+
+    int SupportPenaltyCondition::Check(const ProcessInfo& rCurrentProcessInfo) const

```

Comment:
is there a reason why the check function has return type ìnt`

Reply:
int leaves more options than bool. 0 is the pass here. 1 is if errors have happened. I reckon there could be more codes, but i am not sure if this is caught somewhere...